addOns/pscanrulesAlpha/src/main/java/org/zaproxy/zap/extension/pscanrulesAlpha/JSFunctionPassiveScanner.java

```diff
@@ -0,0 +1,203 @@
+            // Check the scripts in HTML
+            Element el;
+            int offset = 0;
+            while ((el = source.getNextElement(offset, HTMLElementName.SCRIPT)) != null) {
+                String elStr = el.toString();
+                for (Pattern pattern : patterns) {
+                    if (pattern.matcher(elStr).find()) {
+                        evidence.append(elStr);
+                        break; // Only need to record this script once
+                    }
+                }
+                offset = el.getEnd();
+            }
+        } else if (msg.getResponseHeader().isJavaScript()) {
+            // Raw search on response body
+            String content = msg.getResponseBody().toString();
+            for (Pattern pattern : patterns) {
+                if (pattern.matcher(content).find()) {
+                    evidence.append(pattern);
+                    break; // Only need to record one instance of vulnerability
+                }
+            }
+        }
+        if (evidence.length() > 0) {
+            this.raiseAlert(msg, id, evidence.toString());
+        }
+    }
+
+    private void raiseAlert(HttpMessage msg, int id, String evidence) {
+        Alert alert = new Alert(getPluginId(), Alert.RISK_LOW, Alert.CONFIDENCE_LOW, getName());
+        alert.setDetail(
+                this.getDescription(),
+                msg.getRequestHeader().getURI().toString(),
+                "", // Param, not relevant for this vulnerability
+                "", // Attack, not relevant for passive vulnerabilities
+                "", // Other info not specified in message bundle
+                this.getSolution(),
+                this.getReference(),
+                evidence,
+                749, // CWE-749: Exposed Dangerous Method or Function
+                0, // WASC Id - return 0 if no relevant one
+                msg);
+
+        parent.raiseAlert(id, alert);
+    }
+
+    private static void createPatterns() {
+        patterns = new HashSet<>();
+        defaultPatterns = new HashSet<>();
+        try {
+            File f =
+                    new File(
+                            Constant.getZapHome()
+                                    + File.separator
+                                    + FUNC_LIST_DIR
+                                    + File.separator
+                                    + FUNC_LIST_FILE);
+            if (!f.exists()) {
+                throw new IOException("Couldn't find resource: " + f.getAbsolutePath());
+            }
+            try (BufferedReader reader = new BufferedReader(new FileReader(f))) {
+                String line;
+                while ((line = reader.readLine()) != null) {
+                    line = line.trim();
+                    if (!line.startsWith("#") && line.length() > 0) {
+                        defaultPatterns.add(
+                                Pattern.compile("\\b" + line + "\\b", Pattern.CASE_INSENSITIVE));
```

Comment:
- This should be extracted to a method, same for file and payloads.
 - The `line` needs to be quoted (`Pattern#quote`) to handle special regular expressions chars (worth adding a test for the file and custom payloads).

Reply:
During testing I noticed that it works so long as the special characters are not at the start or the end of the function names. For example, `b$dFunction` as an input pattern will match `@b$dFunction()` but `$badFunction` will not match `$badFunction()`. I'm thinking it should be fine as long as the behaviour is documented in the help page?

Reply:
Sounds good.